src/ParallelAlgorithms/Events/ObserveAtExtremum.hpp

```diff
@@ -0,0 +1,412 @@
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "Domain/Structure/ElementId.hpp"
+#include "Domain/Tags.hpp"
+#include "IO/Observer/GetSectionObservationKey.hpp"
+#include "IO/Observer/Helpers.hpp"
+#include "IO/Observer/ObservationId.hpp"
+#include "IO/Observer/ObserverComponent.hpp"
+#include "IO/Observer/ReductionActions.hpp"
+#include "IO/Observer/TypeOfObservation.hpp"
+#include "NumericalAlgorithms/Spectral/Mesh.hpp"
+#include "Options/Options.hpp"
+#include "Parallel/ArrayIndex.hpp"
+#include "Parallel/CharmPupable.hpp"
+#include "Parallel/GlobalCache.hpp"
+#include "Parallel/Invoke.hpp"
+#include "Parallel/Local.hpp"
+#include "Parallel/Reduction.hpp"
+#include "ParallelAlgorithms/Events/Tags.hpp"
+#include "ParallelAlgorithms/EventsAndTriggers/Event.hpp"
+#include "Utilities/ErrorHandling/Assert.hpp"
+#include "Utilities/ErrorHandling/Error.hpp"
+#include "Utilities/Functional.hpp"
+#include "Utilities/Numeric.hpp"
+#include "Utilities/OptionalHelpers.hpp"
+#include "Utilities/TMPL.hpp"
+
+namespace detail {
+
+// Functional called by the reduction opearation: given 2 vectors,
+// return the vector with the largest first component
+struct MaxOfFirstElement {
+  const std::vector<double>& operator()(
+      const std::vector<double>& first, const std::vector<double>& second) {
+    if (UNLIKELY((first.size() != second.size()) or first.size() == 0)) {
+      ERROR("Comparing vectors of different sizes in MaxOfFirstElement");
+    }
+    return (first[0] > second[0]) ? first : second;
+  }
+};
+
+}  // namespace detail
+
+namespace Events {
+/// @{
+/*!
+ * \brief Find the extremum of a Scalar<DataVector> over
+ * all elements, as well as the value of other functions
+ * at the location of that extremum (the reduction returns
+ * the minimum multiplied by (-1) for minima)
+ *
+ * \par Array sections
+ * This event supports sections (see `Parallel::Section`). Set the
+ * `ArraySectionIdTag` template parameter to split up observations into subsets
+ * of elements. The `observers::Tags::ObservationKey<ArraySectionIdTag>` must be
+ * available in the DataBox. It identifies the section and is used as a suffix
+ * for the path in the output file.
+ */
+template <typename ObservationValueTag, typename ObservableTensorTagsList,
+          typename NonTensorComputeTagsList = tmpl::list<>,
+          typename ArraySectionIdTag = void>
+class ObserveAtExtremum;
+
+template <typename ObservationValueTag, typename... ObservableTensorTags,
+          typename... NonTensorComputeTags, typename ArraySectionIdTag>
+class ObserveAtExtremum<ObservationValueTag,
+                        tmpl::list<ObservableTensorTags...>,
+                        tmpl::list<NonTensorComputeTags...>, ArraySectionIdTag>
+    : public Event {
+ private:
+  /// Reduction data will contain the time, and either the maximum
+  /// or the minimum of a function
+  using ReductionData = Parallel::ReductionData<
```

Comment:
[optional] If you want to avoid the "negative of min" thing, I think you could template this on the functor (`funcl::Min` or `funcl::Max`) and choose which to send at runtime.

Reply:
I tried to do this, let me kwow if that's what you had in mind.

Reply:
That is exactly what I was thinking.